Add a way for WKWebView clients to simulate a click over visible text in the page

whsieh opened the pull request:
#### e004985d096c19792336fb1ef1fd9f0ce825e02a
<pre>
Add a way for WKWebView clients to simulate a click over visible text in the page
<a href="https://bugs.webkit.org/show_bug.cgi?id=276613">https://bugs.webkit.org/show_bug.cgi?id=276613</a>
<a href="https://rdar.apple.com/131758473">rdar://131758473</a>

Reviewed by Ryosuke Niwa.

Add a way to simulate mouse events over clickable elements on the webpage, based on matches of found
text. See below for more details.

Tests:  SimulateClickOverText.ClickTargets
        SimulateClickOverText.ClickTargetsAfterScrolling

* Source/WebCore/dom/BoundaryPoint.cpp:
(WebCore::treeOrderInternal):
(WebCore::treeOrder):

Export only `treeOrder&lt;ComposedTree&gt;(const BoundaryPoint&amp;, const BoundaryPoint&amp;)`.

(WebCore::treeOrder&lt;ComposedTree&gt;):
* Source/WebCore/dom/BoundaryPoint.h:
* Source/WebCore/dom/Node.h:

Add `WEBCORE_EXPORT`s to several methods and helper functions.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _pauseNowPlayingMediaSession:]):
(-[WKWebView _simulateClickOverFirstMatchingTextInViewportWithUserInteraction:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::simulateClickOverFirstMatchingTextInViewportWithUserInteraction):

Add plumbing from `WKWebView` -&gt; `WebPageProxy` -&gt; `WebPage`.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::simulateClickOverFirstMatchingTextInViewportWithUserInteraction):

Implement the main heuristic here. At a high level, this is how it works:

1.  Use `findPlainText` to search for the target text in the main document.
2.  Filter out any matches that aren&apos;t within links, or otherwise clickable containers by walking up
    the composed tree.
3.  Compute the first text rect (in content coordinates) for each matching range; filter out any
    matches that are outside of the visual viewport.
4.  Perform one final hit-test to filter out potential targets that are obscured.
5.  If there are 2 or more candidates after the above filtering, exit early and don&apos;t attempt to
    click anything (we may need a more sophisticated strategy to deal with ambiguous targets in the
    future).
6.  Otherwise, if there is exactly one match, create and send synthetic click events.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SimulateClickOverText.mm: Added.
(-[WKWebView simulateClickOverText:]):
(TestWebKitAPI::TEST(SimulateClickOverText, ClickTargets)):
(TestWebKitAPI::TEST(SimulateClickOverText, ClickTargetsAfterScrolling)):

Add a couple of API tests to exercise the new functionality by verifying that mouse and click events
are dispatched on several types of visible (and clickable) DOM elements when simulating clicks based
on found text.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/click-targets.html: Added.

Canonical link: <a href="https://commits.webkit.org/280991@main">https://commits.webkit.org/280991@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9aa5f93f8b6705277914b2778291bda2851a6c83

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58237 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37565 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10720 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61862 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8682 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60366 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45201 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8876 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/47196 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6209 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60268 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35217 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50354 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28031 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31980 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7651 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7686 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53934 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7920 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63567 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2151 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7981 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54513 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2159 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50366 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54572 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/12878 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1843 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8692 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33394 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34480 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35564 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34225 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -9860,6 +9862,120 @@ void WebPage::hasActiveNowPlayingSessionChanged(bool hasActiveNowPlayingSession)
+            FindOption::DoNotSetSelection,
+        });
+
+        if (range.collapsed())
```
Should we assert that targetText is not empty?

whsieh replied:
Good point! Since an empty `targetText` is a misuse of the API, I'll also pair a debug assertion here, with an `NSException` in the `WKWebView` method.

rniwa commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -9860,6 +9862,120 @@ void WebPage::hasActiveNowPlayingSessionChanged(bool hasActiveNowPlayingSession)
+
+        searchRange.start = range.end;
+
+        RefPtr<HTMLElement> target;
```
Why not use lambda?
```
RefPtr<HTMLElement> target = [&] {
    ...
}();
```

whsieh replied:
True, slightly cleaner that way.

rniwa commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -9860,6 +9862,120 @@ void WebPage::hasActiveNowPlayingSessionChanged(bool hasActiveNowPlayingSession)
+        return completion(false);
+    }
+
+    if (candidates.size() >= 2) {
```
I'd say `> 1` instead.

rniwa commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -9860,6 +9862,120 @@ void WebPage::hasActiveNowPlayingSessionChanged(bool hasActiveNowPlayingSession)
     send(Messages::WebPageProxy::HasActiveNowPlayingSessionChanged(hasActiveNowPlayingSession));
 }
 
+void WebPage::simulateClickOverText(const String& targetText, CompletionHandler<void(bool)>&& completion)
```
I'd call this something like `simulateClickOverFirstMatchingTextInViewport` or something

whsieh replied:
Makes sense! I'm renaming this to `simulateClickOverFirstMatchingTextInViewportWithUserInteraction` (to also clarify the `m_userIsInteracting` bit below).

rniwa commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -9860,6 +9862,120 @@ void WebPage::hasActiveNowPlayingSessionChanged(bool hasActiveNowPlayingSession)
+
+    auto& [target, location] = candidates.first();
+
+    SetForScope userIsInteractingChange { m_userIsInteracting, true };
```
It's rather subtle that this function sets this boolean to true.
Maybe this should also be reflected in the function name?

whsieh committed:
Add a way for WKWebView clients to simulate a click over visible text in the page
https://bugs.webkit.org/show_bug.cgi?id=276613
rdar://131758473

Reviewed by Ryosuke Niwa.

Add a way to simulate mouse events over clickable elements on the webpage, based on matches of found
text. See below for more details.

Tests:  SimulateClickOverText.ClickTargets
        SimulateClickOverText.ClickTargetsAfterScrolling

* Source/WebCore/dom/BoundaryPoint.cpp:
(WebCore::treeOrderInternal):
(WebCore::treeOrder):

Export only `treeOrder<ComposedTree>(const BoundaryPoint&, const BoundaryPoint&)`.

(WebCore::treeOrder<ComposedTree>):
* Source/WebCore/dom/BoundaryPoint.h:
* Source/WebCore/dom/Node.h:

Add `WEBCORE_EXPORT`s to several methods and helper functions.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _pauseNowPlayingMediaSession:]):
(-[WKWebView _simulateClickOverFirstMatchingTextInViewportWithUserInteraction:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::simulateClickOverFirstMatchingTextInViewportWithUserInteraction):

Add plumbing from `WKWebView` -> `WebPageProxy` -> `WebPage`.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::simulateClickOverFirstMatchingTextInViewportWithUserInteraction):

Implement the main heuristic here. At a high level, this is how it works:

1.  Use `findPlainText` to search for the target text in the main document.
2.  Filter out any matches that aren't within links, or otherwise clickable containers by walking up
    the composed tree.
3.  Compute the first text rect (in content coordinates) for each matching range; filter out any
    matches that are outside of the visual viewport.
4.  Perform one final hit-test to filter out potential targets that are obscured.
5.  If there are 2 or more candidates after the above filtering, exit early and don't attempt to
    click anything (we may need a more sophisticated strategy to deal with ambiguous targets in the
    future).
6.  Otherwise, if there is exactly one match, create and send synthetic click events.

* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SimulateClickOverText.mm: Added.
(-[WKWebView simulateClickOverText:]):
(TestWebKitAPI::TEST(SimulateClickOverText, ClickTargets)):
(TestWebKitAPI::TEST(SimulateClickOverText, ClickTargetsAfterScrolling)):

Add a couple of API tests to exercise the new functionality by verifying that mouse and click events
are dispatched on several types of visible (and clickable) DOM elements when simulating clicks based
on found text.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/click-targets.html: Added.

Canonical link: https://commits.webkit.org/280991@main
